src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,221 @@
+     * @param {string} userId - Optional userId of constant.
+     * @param {any} value - Value of constant.
+     * @returns {any} Returns the created constant port.
+     */
+    addConst: function (type, userId, value) {
+        var name = userId ? 'const_' + type + '_' + userId : '';
+        if (userId && this._addedNamedInputs[name]) {
+            console.error('pc.ShaderGraphBuilder#addConst: userId ' + userId + ' already used!');
+            return null;
+        }
+        if (!value) {
+            console.error('pc.ShaderGraphBuilder#addConst:' + (userId ? '' : type) + ' constant ' + (userId || '') + ' has no value!');
+            return null;
+        }
+
+        var ioPort = this._graph.addConstant(type, userId, value);
+
+        if (userId) this._addedNamedInputs[name] = ioPort;
+
+        return ioPort;
+    },
+    _validateArg: function (arg, type, name, node, err) {
+        if (arg.type) {
+            if (!this._addedNamedInputs[arg.name] && !arg.name.startsWith('const_') ) {
+                console.error(err + ": invalid input param: " + arg.name);
+                return false;
+            }
+            if (arg.type !== type) {
+                console.error(err + ": input param type mismatch: " + arg.name + "." + type + " != " + node + "." + name + "." + type );
+                return false;
+            }
+        } else {
+            if (typeof(arg) === 'number') {
+                arg = { node: arg, port: 'ret' };
+            }
+            var core = this._addedNodes[arg.node];
+            if (!core) {
+                console.error(err + ": invalid node index: " + arg.node);
+                return false;
+            }
+            var port = core.getIoPortByName(arg.port);
+            if (!port) {
+                console.error(err + ": invalid output port: " + core.name + "." + arg.port);
+                return false;
+            }
+            if (port.type !== type) {
+                console.error(err + ": port type mismatch: " + core.name + "." + port.name + "." + port.type + " != " + node + "." + name + "." + type );
+                return false;
+            }
+        }
+        return true;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description Creates and adds a core node to the shader graph and connects up inputs.
+     * @param {string} name - Name of the core node.
+     * @param {...*} arguments - First argument is name string, then 0-N output node ports or params to
+     * connect to new core node inputs - number and port type of arguments must match.
+     * @returns {number} Returns the created node id.
+     * @example
+     * var node0 = builder.addNode('uv0');
+     * var node1 = builder.addNode('texSample', param0, node0 );
+     * var node2 = builder.addNode('splitVec4', node1 );
+     * var node3 = builder.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {

```

Comment:
I wonder if this should be changed to take an array of ports to connect to inputs, instead of just list of arguments.

currently we call it
`var emissiveCalcNode = builder.addNode('mul3', emissiveColConst, {node:emissiveSampleNode, port:'rgb'});`

but we could change it to 
```
addNode: function (name, inputs)

// and call it as 
var emissiveCalcNode = builder.addNode('mul3', 
    [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ]
 );
```

this would make it easier to read and add additional parameters at a later stage, for example:

```
addNode: function (name, inputs, precision)

// and call it as 
var emissiveCalcNode = builder.addNode('mul3', [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ] );

var emissiveCalcNode = builder.addNode('mul3', 
    [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ],
    pc.PRECISION_HIGH,
    ...
 );

```

It would also make it easier to do meaningful type checking using typescript.

Reply:
Thanks for this suggestion - I have done this as part of the Core Node Generators WIP commit.